File: hush-serve/hush/serve/jobs.py
"""In-memory background job store."""

import asyncio
import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, Optional


class JobStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class Job:
    id: str
    endpoint_path: str
    inputs: Dict[str, Any]
    status: JobStatus = JobStatus.PENDING
    result: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    created_at: float = field(default_factory=time.time)
    completed_at: Optional[float] = None


class JobStore:
    """Thread-safe in-memory job store with TTL cleanup."""

    def __init__(self, ttl_seconds: int = 3600):
        self._jobs: Dict[str, Job] = {}
        self._lock = asyncio.Lock()
        self._ttl = ttl_seconds

    async def create(self, endpoint_path: str, inputs: dict) -> Job:
        async with self._lock:
            job = Job(id=str(uuid.uuid4()), endpoint_path=endpoint_path, inputs=inputs)
            self._jobs[job.id] = job
            self._cleanup_expired()
            return job

    async def get(self, job_id: str) -> Optional[Job]:
        async with self._lock:
            return self._jobs.get(job_id)

    async def update(self, job_id: str, **kwargs) -> None:
        async with self._lock:
            job = self._jobs.get(job_id)
            if job:
                for k, v in kwargs.items():
                    setattr(job, k, v)

    async def list_jobs(self, endpoint_path: Optional[str] = None) -> list:
        async with self._lock:
            jobs = list(self._jobs.values())
            if endpoint_path:
                jobs = [j for j in jobs if j.endpoint_path == endpoint_path]
            return jobs

    def _cleanup_expired(self):
        now = time.time()
        expired = [
            jid
            for jid, job in self._jobs.items()
            if job.completed_at and (now - job.completed_at) > self._ttl
        ]
        for jid in expired:
            del self._jobs[jid]

File: rush-serve/src/jobs.rs
//! In-memory job store for background job execution.

use std::sync::Mutex;
use std::collections::HashMap;

use chrono::{DateTime, Utc};
use serde::Serialize;
use serde_json::Value;

#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum JobStatus {
    Pending,
    Running,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize)]
pub struct Job {
    pub id: String,
    pub status: JobStatus,
    pub result: Option<Value>,
    pub error: Option<String>,
    pub created_at: DateTime<Utc>,
    pub completed_at: Option<DateTime<Utc>>,
}

pub struct JobStore {
    jobs: Mutex<HashMap<String, Job>>,
}

impl JobStore {
    pub fn new() -> Self {
        JobStore {
            jobs: Mutex::new(HashMap::new()),
        }
    }

    pub fn create(&self, id: String) -> Job {
        let job = Job {
            id: id.clone(),
            status: JobStatus::Pending,
            result: None,
            error: None,
            created_at: Utc::now(),
            completed_at: None,
        };
        self.jobs.lock().unwrap().insert(id, job.clone());
        job
    }

    pub fn set_running(&self, id: &str) {
        if let Some(job) = self.jobs.lock().unwrap().get_mut(id) {
            job.status = JobStatus::Running;
        }
    }

    pub fn set_completed(&self, id: &str, result: Value) {
        if let Some(job) = self.jobs.lock().unwrap().get_mut(id) {
            job.status = JobStatus::Completed;
            job.result = Some(result);
            job.completed_at = Some(Utc::now());
        }
    }

    pub fn set_failed(&self, id: &str, error: String) {
        if let Some(job) = self.jobs.lock().unwrap().get_mut(id) {
            job.status = JobStatus::Failed;
            job.error = Some(error);
            job.completed_at = Some(Utc::now());
        }
    }

    pub fn get(&self, id: &str) -> Option<Job> {
        self.jobs.lock().unwrap().get(id).cloned()
    }

    /// Remove completed/failed jobs older than `ttl_seconds`.
    pub fn cleanup(&self, ttl_seconds: i64) {
        let cutoff = Utc::now() - chrono::Duration::seconds(ttl_seconds);
        self.jobs.lock().unwrap().retain(|_, job| {
            if matches!(job.status, JobStatus::Completed | JobStatus::Failed) {
                job.completed_at
                    .map(|t| t > cutoff)
                    .unwrap_or(true)
            } else {
                true
            }
        });
    }
}
